lib/stages/HFBTranspose.cpp

```diff
@@ -118,18 +119,31 @@ bool HFBTranspose::get_dataset_state(dset_id_t ds_id) {
 
     times = tstate->get_times();
     beams = bstate->get_beams();
-    sub_freqs = sfstate->get_subfreqs();
 
     // unzip the vector of pairs in freqState
     for (auto& [id, freq] : fstate->get_freqs()) {
         (void)id;
         freqs.push_back(freq);
     }
 
+    num_subfreq = sfstate->get_subfreqs().size();
     num_time = times.size();
     num_freq = freqs.size();
     num_beams = beams.size();
-    num_subfreq = sub_freqs.size();
+
+    // Populate the sub-frequency index map
+    // Each sub-frequency is the difference from the centre of the sub-band
+    auto& tel = Telescope::instance();
+
+    sub_freqs.resize(num_subfreq);
+
+    const double freq_width = tel.freq_width(0);
+    double freq_diff = 0.5 * freq_width;
+    double freq_inc = freq_width / num_subfreq;
+    for (auto& sf : sub_freqs) {
+        sf = freq_diff;
+        freq_diff -= freq_inc;
```

Comment:
You should try and determine the sign of this (and the start frequency) from the Telescope object, as it depends on the F-engine sampling scheme. I think a decent way of doing this is to use tel::to_freq to determine if the frequencies are increasing or decreasing to determine if you are in an odd or even Nyquist zone and set the sign appropriately.

Maybe @andrerenard can comment about whether this is worth it, but I think that should be correct for most telescopes?

Reply:
So for some upcoming telescopes we will be working in the first and second Nyquist zones.  So yes, I'd suggest using the telescope functions for this.

Reply:
I wounder if we should add an option to `Telescope` or `ICETelescope` to expose the Nyquist zone?  It's a bit odd to have to infer that from the direction of the frequency bin to frequency conversion.

Reply:
@andrerenard, @jrs65 I've added the Nyquist zone to the `ICETelescope` and used it to determine the order of frequencies. Is this what you had in mind?